### PR TITLE
Conversand adapter

### PR DESCRIPTION
**Pull request by pycnvr**
Please let us know the additional fixes needed.  Thank you!

### CODE REVIEW

File: conversant-htb.js [hunk 1]
```diff
@@ -150,13 +265,22 @@ function ConversantHtb(configs) {
...
         return {
             url: baseUrl,
             data: queryObj,
-            callbackId: callbackId
+            callbackId: queryObj.id,
+            
+            /* Signal a POST request and the content type */
+            networkParamOverrides: {
+            	method: 'POST',
+            	contentType: 'application/json'
```
Comment (ix-certification) [inline conversant-htb.js]: Please remove this line 
`contentType: 'application/json'`

File: spec/generateRequestObj.spec.js [hunk 2]
```diff
@@ -92,17 +211,28 @@ describe('generateRequestObj', function () {
...
                         minLength: 1
+                    },
+                    networkParamOverrides: {
+                    	type: 'object',
+                    	properties: {
+                    		method: {
+                    			type: 'string',
+                    			eq: 'POST'
+                    		},
+                    		contentType: {
```
Comment (ix-certification) [inline spec/generateRequestObj.spec.js]: Please remove the contentType validator as it is longer needed as it was removed in the main file.

File: conversant-htb.js [hunk 3]
```diff
@@ -90,8 +204,9 @@ function ConversantHtb(configs) {
      */
     function __generateRequestObj(returnParcels) {
         var queryObj = {};
-        var baseUrl = Browser.getProtocol() + '';
-        var callbackId = System.generateUniqueId();
+        
+        /* todo : specify length and format for cache buster */
+        var baseUrl = Browser.getProtocol() + 'media.msg.dotomi.com/s2s/header/24?cb=' + System.generateUniqueId();
```
Comment (ix-certification) [inline conversant-htb.js]: Please update such that there are double slashes

'//media.msg.dotomi.com/s2s/header/24?cb='

File: conversant-htb-utils.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+'use strict';
```
Comment (ix-certification) [inline conversant-htb-utils.js]: Some refactoring will need to be done. Please remove this file and instead we recommend the following:

1. Move the navigator function in to the main file: conversant-htb-validator.js
2. Utilize `Browser.topWindow` instead for the window

File: conversant-htb.js (hunk 3 above)
Comment (ix-certification) [reply conversant-htb.js]: In addition, please make the following changes to the mock library file for the unit test.

https://github.com/indexexchange/conversant-adapter/blob/d93f702a4b92cc846b970e2037ce720c31091ec9/spec/support/libraryStubData.js#L11-L27

line 13: return 'http://'; = > return 'http:'; (reason why '//' change is needed)
line 19: return 'localhost'; => return 'http://localhost';
line 22: return 'desktop'; => return 'Mozilla/5.0 (Windows; U; Windows NT 6.1; rv:2.2) Gecko/20110201';
Thanks

**Review by ix-certification — CHANGES_REQUESTED** [review]
Please make the recommended changes from the comments. 

In addition, one more change: Please update the callback type from `ID` to `NONE`

https://github.com/indexexchange/conversant-adapter/blob/430f877cb0a9ca0a69ded043eb22f7378b98633a/conversant-htb.js#L546

Please let us know if there are any issues or questions

File: conversant-htb-utils.js (hunk 4 above)
Comment (pycnvr) [reply conversant-htb-utils.js]: Thanks!  Having Browser.topWindow would make things a lot easier.  It doesn't seem to be documented, though.  Could you verify it's just a call with no argument?

Comment (pycnvr) [reply conversant-htb-utils.js]: Oh topWindow is a property.  Perhaps I can add it to the stub as well.

**Comment by pycnvr** [comment]
Thanks for the feedbacks.  Please take a look at the new change.

**Comment by ix-certification** [comment]
Hi Conversant Team,

Thank you for making those changes. Can you please confirm whether the mock config is intended? https://github.com/indexexchange/conversant-adapter/blob/development-v2/spec/support/mockPartnerConfig.json 

Based on the current mock config, the htSlot1 (a representation of an ad slot on a page) will be mapped to xSlot 1 and 2,  which means htSlot1 is expected to have 3 sizes  [300, 250], [180, 150], and [300, 600].

**Comment by pycnvr** [comment]
Yes, while unrealistic, it's meant to test combinations of requests.  It can definitely be changed if there is a different expectation.  Thanks.

**Comment by ix-certification** [comment]
Hi Conversant Team,

Can you please confirm whether your endpoint supports CROS? We run into this error 
`XMLHttpRequest cannot load https://media.msg.dotomi.com/s2s/header/24?cb=sLciWUvE. No 'Access-Control-Allow-Origin' header is present on the requested resource.` 
while testing locally. 

Thanks

**Comment by pycnvr** [comment]
We should.  I ran a test from my local machine to the same endpoint and do see Access-Control-Allow-Origin with my host name returned in the response.  The sandbox page that your team previously setup is also getting the CORS header back.

**Comment by pycnvr** [comment]
Would you mind sending a copy of the request including the url, body, and headers if possible?

**Comment by ix-certification** [comment]
Hi 
Is this good on your end? 
```
CURL
curl 'http://media.msg.dotomi.com/s2s/header/24?cb=woxiMShO' -H 'Pragma: no-cache' -H 'Origin: http://sandbox.ht.indexexchange.com' -H 'Accept-Encoding: gzip, deflate' -H 'Accept-Language: en-CA,en;q=0.8,zh-CN;q=0.6,zh;q=0.4' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.113 Safari/537.36' -H 'Content-Type: application/x-www-form-urlencoded; charset=UTF-8' -H 'Accept: */*' -H 'Cache-Control: no-cache' -H 'Referer: http://sandbox.ht.indexexchange.com/ht/adapter-test-page.html' -H 'Cookie: DotomiNet=2$Bz4ublZzRXRDAmBdAAp6WgNHITlXewBTVxo9WlFdMzwicgo%3D; DotomiUser=112404053078117060$3$42' -H 'Connection: keep-alive' --data '{"id":"yqtqz0Y1","imp":[{"id":"1_0","secure":0,"displaymanager":"40834-index-client","displaymanagerver":"0.0.1","tagid":"54321","banner":{"format":[{"w":300,"h":250}]}}],"site":{"id":"123456","mobile":0,"page":"http://sandbox.ht.indexexchange.com/ht/adapter-test-page.html"},"device":{"h":800,"w":1280,"dnt":0,"language":"en-US","make":"Google Inc.","ua":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.113 Safari/537.36"},"at":1}' --compressed


Request Headers
POST /s2s/header/24?cb=woxiMShO HTTP/1.1
Host: media.msg.dotomi.com
Connection: keep-alive
Content-Length: 488
Pragma: no-cache
Cache-Control: no-cache
Origin: http://sandbox.ht.indexexchange.com
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.113 Safari/537.36
Content-Type: application/x-www-form-urlencoded; charset=UTF-8
Accept: */*
Referer: http://sandbox.ht.indexexchange.com/ht/adapter-test-page.html
Accept-Encoding: gzip, deflate
Accept-Language: en-CA,en;q=0.8,zh-CN;q=0.6,zh;q=0.4
Cookie: DotomiNet=2$Bz4ublZzRXRDAmBdAAp6WgNHITlXewBTVxo9WlFdMzwicgo%3D; DotomiUser=112404053078117060$3$42


Response Headers
HTTP/1.1 204 No Content
Date: Fri, 22 Sep 2017 20:13:56 GMT
Server: Apache/2.2.31 (Unix) mod_apreq2-20090110/2.8.0
Content-Length: 0
Content-Type: text/plain
P3P: policyref="/w3c/p3p.xml", CP="NOI NID DEVo TAIo PSAo HISo OTPo OUR DELo BUS COM NAV INT DSP COR"
Keep-Alive: timeout=5, max=100000
Connection: Keep-Alive


```

**Comment by pycnvr** [comment]
This looks ok.  The only thing preventing it from getting a bid is the site id.  Try 108060 like the one defined in mockPartnerConfig.json should work.

**Comment by ix-certification** [comment]
Hi Conversant Team,

Can you please assign the value `''` to the variable `bidDealId` in this line?https://github.com/indexexchange/conversant-adapter/blob/3b5333b1ba44b176e3bcc89f70b6cc8be71276ec/conversant-htb.js#L410

Thanks

**Comment by pycnvr** [comment]
Done.  Thanks.

File: conversant-htb.js [hunk 5]
```diff
@@ -174,6 +296,12 @@ function ConversantHtb(configs) {
     function adResponseCallback(adResponse) {
         /* get callbackId from adResponse here */
         var callbackId = 0;
+        
+        if (adResponse.hasOwnProperty('id'))
+        	callbackId = adResponse.id;
```
Comment (ix-certification) [inline conversant-htb.js]: Can you please add the `{}` back?

File: conversant-htb.js [hunk 6]
```diff
@@ -72,11 +72,124 @@ function ConversantHtb(configs) {
...
+    	// as an alternative approach.
+    	
+    	for (var i = 0; i < returnParcels.length; ++i) {
+    		var parcel = returnParcels[i];
+    		var xSlot = parcel.xSlotRef;
+    		var imp = {};
+    		var banner = {};
+    		var id = parcel.xSlotName + '_' +  counter++;
+    		
+    		parcel.cnvrId = id;
```
Comment (ix-certification) [inline conversant-htb.js]: Hi, can you please switch to the requestId approach? The parcels are used across the whole request response cycle and modification to the parcel object can cause  unwanted collisions.

File: conversant-htb.js [hunk 7]
```diff
@@ -249,9 +388,10 @@ function ConversantHtb(configs) {
                  * is usually some sort of placements or inventory codes. Please replace the someCriteria
                  * key to a key that represents the placement in the configuration and in the bid responses.
                  */
-
-                if (unusedReturnParcels[j].someCriteria === bids[i].someCriteria) { // change this
+            	
+                if (unusedReturnParcels[j].cnvrId === bids[i].id) { // change this
```
Comment (ix-certification) [inline conversant-htb.js]: Can these be changed according to the requestid approach as well?

File: conversant-htb.js (hunk 6 above)
Comment (pycnvr) [reply conversant-htb.js]: Certainly.  Just want to double check, are requestIds unique?

**Comment by pycnvr** [comment]
Thanks for the feedback.  Changes uploaded.

**Comment by pycnvr** [comment]
Much appreciated.

**Pull request merged**